lighthouse-core/computed/metrics/lantern-first-contentful-paint.js

```diff
@@ -24,52 +26,109 @@ class LanternFirstContentfulPaint extends LanternMetric {
+    // EvaluateScript task finished before FCP as well.
+    /** @type {Set<string>} */
+    const actuallyRenderBlockingScriptUrls = new Set();
+    const blockingCpuNodeIds = new Set();
+    for (const url of possiblyRenderBlockingScriptUrls) {
+      for (const cpuNode of cpuNodes) {
+        if (cpuNode.isEvaluateScriptFor(new Set([url]))) {
+          actuallyRenderBlockingScriptUrls.add(url);
+          blockingCpuNodeIds.add(cpuNode);
+          break;
+        }
+      }
+    }
+
+    const firstLayout = cpuNodes.find(node => node.didPerformLayout());
+    if (firstLayout) blockingCpuNodeIds.add(firstLayout.id);
+    const firstPaint = cpuNodes.find(node => node.childEvents.some(e => e.name === 'Paint'));
+    if (firstPaint) blockingCpuNodeIds.add(firstPaint.id);
+    const firstParse = cpuNodes.find(node => node.childEvents.some(e => e.name === 'ParseHTML'));
+    if (firstParse) blockingCpuNodeIds.add(firstParse.id);
+
+    if (cpuFilter) cpuNodes.filter(cpuFilter).forEach(node => blockingCpuNodeIds.add(node.id));
+
+    return {possiblyRenderBlockingScriptUrls, actuallyRenderBlockingScriptUrls, blockingCpuNodeIds};
   }
 
   /**
    * @param {Node} dependencyGraph
-   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @param {number} fcpTs
+   * @param {function(NetworkNode):boolean} blockingScriptFilter
+   * @param {function(CPUNode):boolean=} cpuNodeFilter
    * @return {Node}
    */
-  static getPessimisticGraph(dependencyGraph, traceOfTab) {
-    const fcp = traceOfTab.timestamps.firstContentfulPaint;
-    const blockingScriptUrls = LanternMetric.getScriptUrls(dependencyGraph, node => {
-      return node.endTime <= fcp && node.hasRenderBlockingPriority();
-    });
+  static getFirstPaintBasedGraph(dependencyGraph, fcpTs, blockingScriptFilter, cpuNodeFilter) {
+    const {
+      possiblyRenderBlockingScriptUrls,
+      actuallyRenderBlockingScriptUrls,
+      blockingCpuNodeIds,
+    } = this.getBlockingCpuData(dependencyGraph, fcpTs, blockingScriptFilter, cpuNodeFilter);
 
     return dependencyGraph.cloneWithRelationships(node => {
-      if (node.endTime > fcp && !node.isMainDocument()) return false;
-      // Include EvaluateScript tasks for blocking scripts
-      if (node.type === BaseNode.TYPES.CPU) {
-        return node.isEvaluateScriptFor(blockingScriptUrls);
-      }
+      if (node.type === BaseNode.TYPES.NETWORK) {
+        // Exclude all nodes that ended after FCP (except for the main document which we always consider necessary)
+        if (node.endTime > fcpTs && !node.isMainDocument()) return false;
 
-      // Include non-script-initiated network requests with a render-blocking priority
-      return node.hasRenderBlockingPriority();
+        const url = node.record.url;
+        if (possiblyRenderBlockingScriptUrls.has(url))
+          return actuallyRenderBlockingScriptUrls.has(url);
+        return node.hasRenderBlockingPriority();
```

Comment:
why override the check in `hasRenderBlockingPriority` if the url is in `possiblyRenderBlockingScriptUrls` but not in `actuallyRenderBlockingScriptUrls`?

Reply:
added more comments and deciding to restructure this a bit